examples/deepspeed/huggingface/prepare_lm_data.py

```diff
@@ -0,0 +1,173 @@
+                                    num_proc=preprocessing_num_workers,
+                                    remove_columns=column_names,
+                                    load_from_cache_file=not overwrite_cache)
+
+    if tokenizer.model_max_length < max_seq_len:
+        logger.warning("The max_seq_len passed is larger than the maximum length for the model. Using max_seq_len="
+                       f"{tokenizer.model_max_length}.")
+        max_seq_len = tokenizer.model_max_length
+
+    # Main data processing function that will concatenate all texts from our dataset and generate chunks of max_seq_len
+    def group_texts(examples):
+        # Concatenate all texts.
+        concatenated_examples = {k: sum(examples[k], []) for k in examples.keys()}
+        total_length = len(concatenated_examples[list(examples.keys())[0]])
+        # We drop the small remainder. We could add padding if the model supported it instead of this drop.
+        # You can customize this part to your needs
+        total_length = (total_length // max_seq_len) * max_seq_len
+        # Split chunks of max_len
+        result = {key: [text[i: i+max_seq_len] for i in range(0, total_length, max_seq_len)]
+                  for key, text in concatenated_examples.items()}
+        result["labels"] = result["input_ids"].copy()
+        return result
+
+    # Note that with `batched=True`, this map processes 1,000 texts together,
+    # so group_texts throws away a remainder for each of these groups of 1,000 texts.
+    # You can adjust that batch_size here but a higher value might be slower to preprocess.
+    lm_datasets = tokenized_datasets.map(group_texts,
+                                         batched=True,
+                                         batch_size=preprocessing_batch_size,
+                                         num_proc=preprocessing_num_workers,
+                                         load_from_cache_file=not overwrite_cache)
+
+    return lm_datasets
+
+
+def main() -> None:
+    args = get_args()
+
+    tokenizer = transformers.AutoTokenizer.from_pretrained(pretrained_model_name_or_path=args.tokenizer_name,
+                                                           cache_dir=args.tokenizer_cache_dir,
+                                                           revision=args.tokenizer_revision,
+                                                           use_fast=True)
+    raw_ds = load_raw_dataset(args.dataset_name,
+                              args.dataset_config_name,
+                              args.dataset_cache_dir,
+                              args.validation_split_percentage)
+    lm_datasets = build_datasets(raw_ds,
+                                 tokenizer,
+                                 args.preprocessing_num_workers,
+                                 args.preprocessing_batch_size,
+                                 args.max_seq_len,
+                                 args.overwrite_cache)
```

Comment:
These functions are largely taken from the respective functions in the [CLMTrial](https://github.com/coreweave/coreweave_determined/blob/3421d393c7a3a4afe8c561fd88ab7c86d36d32a5/model_hub/examples/huggingface/language-modeling/clm_trial.py). In that example they made the dataset preprocessing a part of the trial which means when you do an experiment with a number of trials (hyperparameter search) it will redo the processing every trial. That's why I ripped it out into a separate script to be run before any determined experiments.